addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/internal/protocol/mqtt/MySensorsMqttConnection.java

```diff
@@ -0,0 +1,230 @@
+     * all together
+     */
+    @Override
+    public boolean establishConnection() {
+
+        boolean ret = false;
+
+        if (myMqtt.getService() != null) {
+
+            in = null;
+            out = null;
+            out = new PipedOutputStream();
+            in = new PipedInputStream();
+            try {
+                in.connect(out);
+                mysConReader = new MySensorsReader(in);
+                mysConWriter = new MySensorsMqttWriter(System.out);
+                // null output rather than stdout?
+                // it should never be written to however, as we have overridden the sendMessage method
+
+                mqttService = myMqtt.getService();
+
+                topicSubscribe = myGatewayConfig.getTopicSubscribe();
+                topicPublish = myGatewayConfig.getTopicPublish();
+                brokerName = myGatewayConfig.getBrokerName();
+
+                mqttConsumer.setTopic(topicSubscribe + "/+/+/+/+/+");
+
+                mqttService.registerMessageConsumer(brokerName, mqttConsumer);
+                mqttService.registerMessageProducer(brokerName, mqttPublisher);
+
+                ret = startReaderWriterThread(mysConReader, mysConWriter);
+
+            } catch (IOException e) {
+                logger.debug(e.toString());
+            }
+        }
+
+        return ret;
+    }
+
+    /**
+     * Cleans up all resources
+     */
+    @Override
+    protected void stopConnection() {
+        in = null;
+        out = null;
+        mqttService.unregisterMessageConsumer(brokerName, mqttConsumer);
+        mqttService.unregisterMessageProducer(brokerName, mqttPublisher);
+    }
+
+    protected class MySensorsMqttWriter extends MySensorsWriter {
+
+        public MySensorsMqttWriter(OutputStream outStream) {
+            super(outStream);
+        }
+
+        @Override
+        protected void sendMessage(String msg) {
+
+            logger.debug("Sending MQTT Message: Topic: {}, Message: {}", topicPublish, msg);
+            try {
+                mqttPublisher.publish(topicPublish, msg);
+            } catch (NullPointerException ne) {
+                logger.debug("Null exception from MQTT transport service, broker unavailable");
+                MySensorsMqttConnection.this.requestDisconnection(true);
+            } catch (Exception e) {
+                logger.debug("Error sending MQTT message: {}", e.toString());
+            }
+
+        }
+
+    }
+
+    /**
+     * Receives messages from MQTT transport, translates them and passes them on to
+     * the MySensors abstract connection
+     */
+    public class MySensorsMqttConsumer implements MqttMessageConsumer {
+
+        private EventPublisher eventPublisher;
+        private String topic;
+
+        @Override
+        public void processMessage(String topic, byte[] payload) {
+
+            logger.debug("MQTT message received. Topic: {}, Message: {}", topic, payload.toString());
```

Comment:
payload.toString() does not work. Use something like:
`String message = new String(payload);`